File: Interview/DS/Sorting/Shell_sort/shell_sort.py
def shell_sort(arr):
    n = len(arr)
    gap = n // 2
    
    while gap > 0:
        # Do a gapped insertion sort for this gap size.
        for i in range(gap, n):
            temp = arr[i]
            j = i
            while j >= gap and arr[j - gap] > temp:
                c1 = j - gap 
                c2 = arr[c1]
                
                arr[j] = arr[j - gap]
                j -= gap
            arr[j] = temp
        gap //= 2

# Example code to run shell sort

arr = [64, 34, 25, 12, 22, 11, 90]
print("Original array:", arr)

shell_sort(arr)
print("Sorted array:", arr)

File: Interview/DS/Sorting/Shell_sort/shell_sort.java
public class ShellSort {
    public static void main(String[] args) {
        int[] arr = {20, 35, -15, 7, 55, 1, -22};
        shellSort(arr);
        for (int i : arr) {
            System.out.println(i);
        }
    }

    /**
     * The main shell sort method that takes an array of integers as its argument.
     * 
     * @param arr the array to be sorted
     */
    public static void shellSort(int[] arr) {
        int gap = arr.length / 2;
        while (gap > 0) {
            for (int i = gap; i < arr.length; i++) {
                int newElement = arr[i];

                int j = i;
                while (j >= gap && arr[j - gap] > newElement) {
                    arr[j] = arr[j - gap];
                    j -= gap;
                }

                arr[j] = newElement;
            }
            gap /= 2;
        }
    }
}
